src/Storages/StorageMerge.cpp

```diff
@@ -1702,6 +1735,58 @@ std::optional<UInt64> StorageMerge::totalRowsOrBytes(F && func) const
     return first_table ? std::nullopt : std::make_optional(total_rows_or_bytes);
 }
 
+void StorageMerge::setTableToWrite(
+    const std::optional<String> & table_to_write_,
+    const String & source_database_name_or_regexp_,
+    bool database_is_regexp_)
+{
+    if (!table_to_write_.has_value())
+    {
+        table_to_write = std::nullopt;
+        return;
+    }
+
+    auto qualified_name = QualifiedTableName::parseFromString(*table_to_write_);
+
+    if (qualified_name.database.empty())
+    {
+        if (database_is_regexp_)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Argument 'table_to_write' must contain database if 'db_name' is regular expression.");
+
+        qualified_name.database = source_database_name_or_regexp_;
+    }
+
+    table_to_write = qualified_name;
+}
+
+SinkToStoragePtr StorageMerge::write(
+    const ASTPtr & query,
+    const StorageMetadataPtr & metadata_snapshot,
+    ContextPtr context_,
+    bool async_insert)
+{
+    if (table_to_write_auto)
+    {
+        table_to_write = std::nullopt;
+        forEachTableName([&](const auto & table_name)
+        {
+            if (!table_to_write.has_value())
+                table_to_write = table_name;
+            else if (table_to_write->getFullName() < table_name.getFullName())
+                table_to_write = table_name;
+        });
+    }
+
+    if (!table_to_write.has_value())
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Method write is not allowed in storage {} without described table to write.", getName());
+
+    context_->getAccess()->checkAccess(AccessType::INSERT, table_to_write->database, table_to_write->table);
```

Comment:
During reading or creation, we skip tables that the user doesn't have `show` access to. Let's check the same above, it would skip some lexographically less names if a user doesn't see them at all

Reply:
By the way - I don't like it, because two users get different result on same SELECT query depends on their access without  explanation, why.

Reply:
Yes, I see, but we already have this behaviour and should be consistent among operations.
At least this logic hides tables from the user, which he is not supposed to know about.
Behaviour like that would expose database content with any grants
```sql
SELECT * FROM merge('secret_database', '.*') -- you don't have access to secret_database.secret_table
```